apps/emqx_durable_storage/src/emqx_ds_tx_conflict_trie.erl

```diff
@@ -0,0 +1,509 @@
+
+-record(conflict_tree, {
+    rotate_every :: pos_integer() | infinity,
+    min_serial,
+    old_max_serial,
+    max_serial,
+    old_trie = gb_trees:empty(),
+    trie = gb_trees:empty()
+}).
+
+-opaque t() :: #conflict_tree{}.
+
+-type conflict_domain() :: [binary() | '#'].
+
+%%================================================================================
+%% API functions
+%%================================================================================
+
+-spec print(t()) -> map().
+print(#conflict_tree{min_serial = Min, max_serial = Max, old_trie = Old, trie = Current}) ->
+    #{
+        range => {Min, Max},
+        old => gb_trees:to_list(Old),
+        current => gb_trees:to_list(Current)
+    }.
+
+%% @doc Translate topic filter to conflict domain. This function
+%% replaces all topic levels following a wildcard (+ or #) with #
+-spec topic_filter_to_conflict_domain(emqx_ds:topic_filter()) -> conflict_domain().
+topic_filter_to_conflict_domain(TF) ->
+    tf2cd(TF, []).
+
+%% @doc Create a new conflict tracking trie.
+%%
+%% @param MinSerial Transaction serial at the time of creation.
+%%
+%% @param RotateEvery Automatically rotate the trie when the tracked
+%% serial interval reaches this value. Atom `infinity' disables
+%% automatic rotation.
+-spec new(emqx_ds_optimistic_tx:serial(), pos_integer() | infinity) -> t().
+new(MinSerial, RotateEvery) when
+    RotateEvery > 0;
+    RotateEvery =:= infinity
+->
+    #conflict_tree{
+        rotate_every = RotateEvery,
+        min_serial = MinSerial,
+        max_serial = MinSerial,
+        old_max_serial = MinSerial
+    }.
+
+%% @doc Add a new conflict domain to the trie.
+-spec push(conflict_domain(), emqx_ds_optimistic_tx:serial(), t()) -> t().
+push(
+    CD,
+    Serial,
+    S = #conflict_tree{min_serial = MinS, max_serial = MaxS}
+) when Serial > MinS, Serial >= MaxS ->
+    #conflict_tree{rotate_every = RotateEvery, old_max_serial = OldMax, trie = Trie} = S,
+    case Serial - OldMax >= RotateEvery of
+        true ->
+            push(CD, Serial, rotate(S));
+        false ->
+            WildcardPrefix = static_prefix(CD),
+            S#conflict_tree{
+                max_serial = Serial,
+                trie = do_push(WildcardPrefix, CD, Serial, Trie)
+            }
+    end.
+
+%% @doc Return `false' if there are no entries with serial **greater**
+%% than `Serial' in the tracked conflict range. Otherwise, return
+%% `true'.
+-spec is_dirty(conflict_domain(), emqx_ds_optimistic_tx:serial(), t()) -> boolean().
+is_dirty(
+    _CD,
+    Serial,
+    #conflict_tree{min_serial = MinS}
+) when Serial < MinS ->
+    %% Serial is out of the tracked conflict range:
+    true;
+is_dirty(
+    CD,
+    Serial,
+    #conflict_tree{trie = Trie, old_trie = OldTrie}
+) ->
+    check_dirty(CD, Serial, Trie) orelse
+        check_dirty(CD, Serial, OldTrie).
+
+%% @doc This function is used to reduce size of the trie by removing old
```

Comment:
please add some info about the practical implication to transactions after rotation happens when older conflict set is dropped.

Reply:
Fixed: https://github.com/emqx/emqx/pull/15208/commits/4cab80ae3e762bdce862e8b7fea8fe72b78c92a9